sympy/polys/matrices/tests/test_fflu.py

```diff
@@ -0,0 +1,302 @@
+        DM([[2, 0, 0], [0, 12, 0], [0, 0, 144]], ZZ),
+        DM([[2, 0, 0], [0, 6, 0], [0, 0, 24]], ZZ)
+
+    ),
+
+    (
+        'rank_deficient_3x3',
+        DM([[1, 2, 3], [2, 4, 6], [3, 6, 9]], ZZ),
+        DM([[1, 0, 0], [0, 1, 0], [0, 0, 1]], ZZ),
+        DM([[1, 0, 0], [2, 1, 0], [3, 0, 1]], ZZ),
+        DM([[1, 0, 0], [0, 0, 0], [0, 0, 0]], ZZ),
+        DM([[1, 2, 3], [0, 0, 0], [0, 0, 0]], ZZ),
+    ),
+
+    (
+        'zz_1x1',
+        DM([[5]], ZZ),
+        DM([[1]], ZZ),
+        DM([[5]], ZZ),
+        DM([[5]], ZZ),
+        DM([[5]], ZZ),
+    ),
+
+    (
+        'zz_nx1_2rows',
+        DM([[81], [54]], ZZ),
+        DM([[1, 0], [0, 1]], ZZ),
+        DM([[81, 0], [54, 81]], ZZ),
+        DM([[81, 0], [0, 81]], ZZ),
+        DM([[81], [0]], ZZ),
+    ),
+
+    (
+        'zz_nx2_3rows',
+        DM([[2, 7], [7, 45], [25, 84]], ZZ),
+        DM([[1, 0, 0], [0, 1, 0], [0, 0, 1]], ZZ),
+        DM([[2, 0, 0], [7, 82, 0], [25, 41, 41]], ZZ),
+        DM([[2, 0, 0], [0, 82, 0], [0, 0, 41]], ZZ),
+        DM([[2, 7], [0, 82], [0, 0]], ZZ),
+    ),
+
+    (
+
+        'zz_1x2',
+        DM([[0, 28]], ZZ),
+        DM([[1]], ZZ),
+        DM([[28]], ZZ),
+        DM([[28]], ZZ),
+        DM([[0, 28]], ZZ)
+    ),
+
+    (
+        'zz_nx3_4rows',
+        DM([[84, 30, 9], [20, 59, 13], [53, 46, 81], [63, 48, 29]], ZZ),
+        DM([[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1]], ZZ),
+        DM([[84, 0, 0, 0], [20, 365904, 0, 0], [53, 303411, 303411, 0], [63, 303411, 303411, 303411]], ZZ),
+        DM([[84, 0, 0, 0], [0, 365904, 0, 0], [0, 0, 1321658316, 0], [0, 0, 0, 303411]], ZZ),
+        DM([[84, 30, 9], [0, 365904, 13], [0, 0, 1321658316], [0, 0, 0]], ZZ),
+    ),
+
+    (
+        'fflu_row_swap',
+        DM([[0, 1, 2], [3, 4, 5], [6, 7, 8]], ZZ),
+        DM([[0, 1, 0], [1, 0, 0], [0, 0, 1]], ZZ),
+        DM([[3, 0, 0], [0, 3, 0], [6, -3, 1]], ZZ),
+        DM([[3, 0, 0], [0, 9, 0], [0, 0, 3]], ZZ),
+        DM([[3, 4, 5], [0, 3, 6], [0, 0, 0]], ZZ)
+    ),
+]
+
+
+def _check_fflu(A, P, L, D, U):
+    A_field = _to_DM(A, P).to_field().to_dense()
+    P_field = P.to_field().to_dense()
+    L_field = L.to_field().to_dense()
+    D_field = D.to_field().to_dense()
+    U_field = U.to_field().to_dense()
+    m, n = A.shape
+    assert P_field.shape == (m, m)
+    assert L_field.shape == (m, m)
+    assert D_field.shape == (m, m)
+    assert U_field.shape == (m, n)
+    assert L_field.is_lower
+    assert D_field.is_diagonal
+    if hasattr(D, 'inv_den'):
+        di, d = D.inv_den()
+        assert P.matmul(A).rmul(d) == L.matmul(di).matmul(U)
+    else:
+        assert L_field.matmul(D_field).matmul(U_field) == P_field.matmul(A_field)
+    assert U_field.is_upper
+
+
+def _to_DM(A, ans):
+    if isinstance(A, DomainMatrix):
+        return A
+    elif isinstance(A, Matrix):
+        return A.to_DM(ans.domain)
+
+    if not (hasattr(A, 'shape') and hasattr(A, 'domain')):
```

Comment:
All the inputs here should have shape and domain. This can just be `return DomainMatrix(A.to_list(), A.shape, A.domain)`.